[SPARK-28148][SQL] Repartition after join is not optimized away

bmarcott opened the pull request:
### What changes were proposed in this pull request?

Extra shuffling was not eliminated after inner joins because they produce PartitioningCollection Partitioning and the current logic only matched on HashPartitioning. 

Nothing was present in EnsureRequirements to eliminate parent sorting (within partitions) which was unnecessary when the same sort order was introduced by sortmergejoin

Copied from jira:
Partitioning & sorting is usually retained after join.
```
spark.conf.set('spark.sql.shuffle.partitions', '42')

df1 = spark.range(5000000, numPartitions=5)
df2 = spark.range(10000000, numPartitions=5)
df3 = spark.range(20000000, numPartitions=5)

# Reuse previous partitions & sort.
df1.join(df2, on='id').join(df3, on='id').explain()
# == Physical Plan ==
# *(8) Project [id#367L]
# +- *(8) SortMergeJoin [id#367L], [id#374L], Inner
#    :- *(5) Project [id#367L]
#    :  +- *(5) SortMergeJoin [id#367L], [id#369L], Inner
#    :     :- *(2) Sort [id#367L ASC NULLS FIRST], false, 0
#    :     :  +- Exchange hashpartitioning(id#367L, 42)
#    :     :     +- *(1) Range (0, 5000000, step=1, splits=5)
#    :     +- *(4) Sort [id#369L ASC NULLS FIRST], false, 0
#    :        +- Exchange hashpartitioning(id#369L, 42)
#    :           +- *(3) Range (0, 10000000, step=1, splits=5)
#    +- *(7) Sort [id#374L ASC NULLS FIRST], false, 0
#       +- Exchange hashpartitioning(id#374L, 42)
#          +- *(6) Range (0, 20000000, step=1, splits=5)
```

However here: Partitions persist through left join, sort doesn't.

```
df1.join(df2, on='id', how='left').repartition('id').sortWithinPartitions('id').explain()
# == Physical Plan ==
# *(5) Sort [id#367L ASC NULLS FIRST], false, 0
# +- *(5) Project [id#367L]
#    +- SortMergeJoin [id#367L], [id#369L], LeftOuter
#       :- *(2) Sort [id#367L ASC NULLS FIRST], false, 0
#       :  +- Exchange hashpartitioning(id#367L, 42)
#       :     +- *(1) Range (0, 5000000, step=1, splits=5)
#       +- *(4) Sort [id#369L ASC NULLS FIRST], false, 0
#          +- Exchange hashpartitioning(id#369L, 42)
#             +- *(3) Range (0, 10000000, step=1, splits=5)
```
Also here: Partitions do not persist though inner join.

```
df1.join(df2, on='id').repartition('id').sortWithinPartitions('id').explain()
# == Physical Plan ==
# *(6) Sort [id#367L ASC NULLS FIRST], false, 0
# +- Exchange hashpartitioning(id#367L, 42)
#    +- *(5) Project [id#367L]
#       +- *(5) SortMergeJoin [id#367L], [id#369L], Inner
#          :- *(2) Sort [id#367L ASC NULLS FIRST], false, 0
#          :  +- Exchange hashpartitioning(id#367L, 42)
#          :     +- *(1) Range (0, 5000000, step=1, splits=5)
#          +- *(4) Sort [id#369L ASC NULLS FIRST], false, 0
#             +- Exchange hashpartitioning(id#369L, 42)
#                +- *(3) Range (0, 10000000, step=1, splits=5)
```

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

dbtsai commented:
Jenkins, ok to test.

SparkQA commented:
**[Test build #116187 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116187/testReport)** for PR 27096 at commit [`ff573e8`](https://github.com/apache/spark/commit/ff573e864c8271d157acd2ae3b62de5aeb03117a).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20980/
Test PASSed.

SparkQA commented:
**[Test build #116187 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116187/testReport)** for PR 27096 at commit [`ff573e8`](https://github.com/apache/spark/commit/ff573e864c8271d157acd2ae3b62de5aeb03117a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116187/
Test PASSed.

dongjoon-hyun commented:
Thank you for making a PR, @bmarcott .

dbtsai commented:
also ping @viirya

viirya reviewed (commented):
Currently `EnsureRequirements` ensures children distribution/ordering satisfies parent's requirement. That said, it checks if a child's distribution/ordering is good for its parent op. It not, it adds shuffle/sort on top of the child.

The cases failed to optimize I've seen so far, are incorrect or mismatched distribution/ordering output/requirement at child/parent sides.

For this part in `EnsureRequirements`:

```scala
def apply(plan: SparkPlan): SparkPlan = plan.transformUp {
  // TODO: remove this after we create a physical operator for `RepartitionByExpression`.
  case operator @ ShuffleExchangeExec(upper: HashPartitioning, child, _) =>
    child.outputPartitioning match {
      case lower: HashPartitioning if upper.semanticEquals(lower) => child
      case _ => operator
    }
}
```

Read from the TODO, looks like it is supposed to have physical `RepartitionByExpression` that can skip shuffle if its child already satisfies the required distribution?

This PR targets a different scenario as the repartition/sort is added explicitly by the user. We have optimization at Optimizer that deals with logical plan, for example to remove Sort after sorted child plan.

So maybe `EnsureRequirements` not a good place to do that optimization? Maybe we need a separate rule for that? cc @cloud-fan

maropu commented:
(The @viirya suggestion soulds resonable to me)

cloud-fan commented:
an issue is that at logical phase we don't know the physical partitioning/sorting info (e.g. SMJ), so we can't optimize the user-specified sort/repartition well at logical phase.

I see 2 options here:
1. run an extra physical rule after `EnsureRequirements`, to remove useless shuffles
2. do as the TODO says, create a new physical node for `RepartitionByExpression` then we can go through the main logic in the `ensureDistributionAndOrdering` method, instead of a small hack in the `apply` method.

bmarcott commented:
Thanks for taking a look!
Yes, the reason it is here is because the shuffle/sorting is introduced by EnsureRequirements itself, which causes the user added sorts/shuffles unnecessary. Yea it felt a little hacky for optimization code to be in a rule called EnsureRequirements.

I'd like someone more familiar with overall planner design to suggest whether I go through with 1st or 2nd option.
For 2nd option, won't I need to create a new physical node for both the repartition and sort, each of which is kinda a dummy physical node which relies on EnsureRequirements to add the necessary sorts/partitioning based on `requiredChildDistribution` and `requiredChildOrdering`

bmarcott commented:
Any more suggestions?
I'm starting to think the new physical node is a bit of a hack as well just to reuse the code in `ensureDistributionAndOrdering`. We wouldn't want this "no op" SparkPlan to show up in the plan right?

cloud-fan commented:
can we try option 1? Seems like we need to do some experiments here. I'm not sure which option is better without seeing the concrete code.

ulysses-you commented:
+1, how about add a rule named like `PruneShuffleAndSort`.
At least include:

1. 
```
// TODO: remove this after we create a physical operator for `RepartitionByExpression`.
case operator @ ShuffleExchangeExec(upper: HashPartitioning, child, _) =>
  child.outputPartitioning match {
    case lower: HashPartitioning if upper.semanticEquals(lower) => child
    case _ => operator
  }
```

2. this pr code

3. [PR-26946](https://github.com/apache/spark/pull/26946)

bmarcott commented:
i'll update soon with the new rule `PruneShuffleAndSort`
I'm iffy on including [PR-26946](https://github.com/apache/spark/pull/26946).
`EnsureRequirements` should probably only introduce shuffles/ordering if it needs to, rather than adding blindly and relying on later rule to prune. Or maybe it is better to include so it'll apply to any case (haven't thought deeply to the cases here)?

bmarcott committed:
create PruneShuffleAndSort physical rule

SparkQA commented:
**[Test build #116740 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116740/testReport)** for PR 27096 at commit [`1916ef9`](https://github.com/apache/spark/commit/1916ef9b679aa963d75512082b092199b97bcb9b).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21515/
Test PASSed.

bmarcott commented:
I started to try including #26946, but it feels a little messy.

Notice that the numPartitions of the previous shuffle are maintained, but the distribution/partitioning is changed:
`case (ShuffleExchangeExec(partitioning, child, _), distribution: OrderedDistribution) =>
        ShuffleExchangeExec(distribution.createPartitioning(partitioning.numPartitions), child)`

If we try to remove this code from EnsureRequirements, it'll create a new shuffle node with `defaultNumPreShufflePartitions`. Does it make sense to have a general rule that if we have a shuffle with range partitioning with another shuffle as child, we eliminate the child but reuse it's numpartitions?

SparkQA commented:
**[Test build #116740 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116740/testReport)** for PR 27096 at commit [`1916ef9`](https://github.com/apache/spark/commit/1916ef9b679aa963d75512082b092199b97bcb9b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class PruneShuffleAndSort() extends Rule[SparkPlan] `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116740/
Test FAILed.

cloud-fan commented:
In general, shuffles are added by Spark and Spark can pick the best # of partitions. However, for user-specified shuffles (df.repartition), the semantic should be respecting the user request and keep the # of partitions.

bmarcott committed:
fix match error

SparkQA commented:
**[Test build #116772 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116772/testReport)** for PR 27096 at commit [`85003be`](https://github.com/apache/spark/commit/85003be5c55e6be49f9b59d49f2cd8df85d779f2).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21546/
Test PASSed.

SparkQA commented:
**[Test build #116772 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116772/testReport)** for PR 27096 at commit [`85003be`](https://github.com/apache/spark/commit/85003be5c55e6be49f9b59d49f2cd8df85d779f2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116772/
Test PASSed.

bmarcott committed:
add test

SparkQA commented:
**[Test build #116815 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116815/testReport)** for PR 27096 at commit [`00b082a`](https://github.com/apache/spark/commit/00b082a1e24d79c8471d4131e136ef8079e9239e).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21588/
Test PASSed.

bmarcott committed:
add doc

SparkQA commented:
**[Test build #116820 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116820/testReport)** for PR 27096 at commit [`2517bea`](https://github.com/apache/spark/commit/2517beae0b2382a9171239d5682364957135318c).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21593/
Test PASSed.

bmarcott commented:
updated with a new sparkplan rule and added a test which makes sure a user's repartition with a different numPartitions would not be eliminated (don't want to change expected numPartitions).

Review when you get a chance :).

SparkQA commented:
**[Test build #116821 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116821/testReport)** for PR 27096 at commit [`9e895bd`](https://github.com/apache/spark/commit/9e895bd16162577196b83add1ef3d99bb4fc0d08).

SparkQA commented:
**[Test build #116821 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116821/testReport)** for PR 27096 at commit [`9e895bd`](https://github.com/apache/spark/commit/9e895bd16162577196b83add1ef3d99bb4fc0d08).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116821/
Test FAILed.

SparkQA commented:
**[Test build #116820 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116820/testReport)** for PR 27096 at commit [`2517bea`](https://github.com/apache/spark/commit/2517beae0b2382a9171239d5682364957135318c).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #116815 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116815/testReport)** for PR 27096 at commit [`00b082a`](https://github.com/apache/spark/commit/00b082a1e24d79c8471d4131e136ef8079e9239e).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116820/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116815/
Test FAILed.

bmarcott commented:
retest this please

bmarcott commented:
do the tests run post merge with master, or do I need to rebase to pick up fix  #27242?

maropu commented:
retest this please

SparkQA commented:
**[Test build #117038 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117038/testReport)** for PR 27096 at commit [`9e895bd`](https://github.com/apache/spark/commit/9e895bd16162577196b83add1ef3d99bb4fc0d08).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21803/
Test PASSed.

SparkQA commented:
**[Test build #117038 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117038/testReport)** for PR 27096 at commit [`9e895bd`](https://github.com/apache/spark/commit/9e895bd16162577196b83add1ef3d99bb4fc0d08).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/117038/
Test PASSed.

bmarcott commented:
rebased

SparkQA commented:
**[Test build #117198 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117198/testReport)** for PR 27096 at commit [`3ecdccb`](https://github.com/apache/spark/commit/3ecdccb29fe79a1f56c8fbff987b2ac69c870cad).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21958/
Test PASSed.

SparkQA commented:
**[Test build #117198 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117198/testReport)** for PR 27096 at commit [`3ecdccb`](https://github.com/apache/spark/commit/3ecdccb29fe79a1f56c8fbff987b2ac69c870cad).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/117198/
Test PASSed.

bmarcott committed:
update doc

bmarcott commented:
any remaining comments?

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21970/
Test PASSed.

SparkQA commented:
**[Test build #117211 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117211/testReport)** for PR 27096 at commit [`bb794db`](https://github.com/apache/spark/commit/bb794db82dcabfc9c9bb2a309fc70566b3da1ac0).

SparkQA commented:
**[Test build #117211 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117211/testReport)** for PR 27096 at commit [`bb794db`](https://github.com/apache/spark/commit/bb794db82dcabfc9c9bb2a309fc70566b3da1ac0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/117211/
Test PASSed.

bmarcott commented:
@cloud-fan and/or others please review when convenient

bmarcott commented:
mind taking another look?
@cloud-fan 
@viirya

dongjoon-hyun commented:
Retest this please.

SparkQA commented:
**[Test build #117679 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117679/testReport)** for PR 27096 at commit [`bb794db`](https://github.com/apache/spark/commit/bb794db82dcabfc9c9bb2a309fc70566b3da1ac0).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/22440/
Test PASSed.

SparkQA commented:
**[Test build #117679 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117679/testReport)** for PR 27096 at commit [`bb794db`](https://github.com/apache/spark/commit/bb794db82dcabfc9c9bb2a309fc70566b3da1ac0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/117679/
Test PASSed.

bmarcott commented:
@cloud-fan @viirya 
bump 😃

maryannxue commented on sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/PruneShuffleAndSort.scala:
```diff
@@ -0,0 +1,52 @@
+    plan.transformUp {
+      case operator @ ShuffleExchangeExec(upper: HashPartitioning, child, _) =>
+        child.outputPartitioning match {
+          case lower: HashPartitioning if upper.semanticEquals(lower) => child
```
Why doesn't this apply to `RangeParititoning`? Is it just we assume `repartition()` would only do `HashPartitioning`?
And what happens if someone does `df1.join(df2, Seq("id"), "left").repartition(100, df("some_other_column")).repartition(20, df1("id"))` or `df1.join(df2, Seq("id"), "left").sortWithinPartition(df1("some_other_column")).sortWithinPartition(df1("id"))` ? We should be able to optimize that out too, right? It would be nice to make this rule more general and cover a wider range of cases.

bmarcott replied:
Thanks for providing feedback.
Let me take a look into your specific examples  and think a little more about it.

bmarcott replied:
This PR focused on fixing removing unnecessary sorting and shuffling after a join, which potentially includes its own ShuffleExchangeExec with HashPartioning. Both cases you mentioned are already optimized properly:  the shuffling on "some_other_column" is removed and all sortWithinPartitions are removed (due to previous optimizations in logical plan, and the optimizations introduced here)

I wouldn't mind generalizing to all Partitioning types of the ShuffleExchangeExec, but I am not sure how to compare two partitioning types for equality. You can see the special case for HashPartitioning in this PR.

bmarcott replied:
@maryannxue ping

bmarcott commented:
Fixed latest conflicts/rebased on latest master

@cloud-fan  could you take a look when you get a chance?

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/26332/
Test PASSed.

SparkQA commented:
**[Test build #121654 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121654/testReport)** for PR 27096 at commit [`9db29e6`](https://github.com/apache/spark/commit/9db29e60cb0f3fc033b8ccf8e414fd9d3eafa257).

SparkQA commented:
**[Test build #121654 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121654/testReport)** for PR 27096 at commit [`9db29e6`](https://github.com/apache/spark/commit/9db29e60cb0f3fc033b8ccf8e414fd9d3eafa257).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/121654/
Test FAILed.

bmarcott commented:
@cloud-fan @viirya could you help review or add a suggested reviewer here?

bmarcott commented:
@cloud-fan @viirya friendly ping
